### DISCUSSION TITLE
Filament Runout and Resume Procedure

### DISCUSSION THREAD

**Discussion by gg48gg** [xblax/flashforge_ad5m_klipper_mod#220] [Q&A]
Hi folks,

I just want to share my experience with what happens when my printer runs out of filament and I attempt to re-supply and resume the print.  I believe there is room for improvement either in my procedure or software or perhaps both.  Please let me know how I can improve my experience with this.

Filament ran out and here is what I experienced.
1) aux part cooling fan is on high
2)  bed is cooled to room temperature, print is at risk for unsticking from the build plate.
3)  screen shows only a single row of options/buttons on the bottom of the screen.  Load, unload, resume, etc.
4)  I press unload, nothing happens and the screen changes to home screen and freezes.
5)  I am able to get to luid (https://ipaddress:4001).
6)  I am presented  with a popup that filament has runout, with potentially options.  None of the options work, all with warning that nozzle temperature is too low.
7)  I dismiss/close the popup window
8)  I set bed temperature to 70
9)  I set nozzle temperature to 220
10)  I wait for temps to stabilize
11)  While waiting, I  try to cut the old filament by removing the tube from the hotend and cutting the filament, but the tool head is parked at the front left corner and it is difficult to get to as the front lip of the machine is partially in the way.  I manage to cut the filament about 1/2 inch above the hotend.
12)  Still unable to use the screen (frozen), I go to fluid and issue the extrude command while running back and forth to the printer to manage filament loading.  I finally manage to get new filament flowing and the nozzle cleaned up.
13)  back in fluidd, I see that the current state is still "paused" and hit the resume button.  I get a message saying that the axes must be homed and coordinates such as 99.x, 99.x, 99.x which do not seem correct.  Resume print fails.
14)  nervous about homing all axes with a 50mm high print on the bed, I home just the X and Y axes successfully via the fluidd interface.  
15)  Since the bed moved down (Z-) during the x and y axes homing, I am nervous and doubting that the print will be able to resume.  I hit resume and the resume was successful.

Unbelievably, this is a better experience than stock setup, since I was never able to get the printer to resume successfully in stock form.  Still, I would love to know if I am doing something wrong, or what I/we can do to make this a better experience.  Thank you so much for your work on this project.  If I am able to provide any diagnostic or log information, I would be happy to do so.  

Thank you!

Printer: Adventurer 5M Pro 
[Adventurer 5M Klipper Mod v00.05-beta](https://github.com/xblax/flashforge_ad5m_klipper_mod/releases/tag/v00.05-beta)

**Comment by xblax** [comment]
Hi @gg48gg

I assume that you run in the 10min idle timeout of the printer. I think if the idle timeout triggers, the steppers and heaters will be disabled, even if waiting for a filament change. In that idle state, then the unload/load buttons obviously don't work. You need to manually set the temperature and home all axis, as you did.

I'm not sure what we can do to improve this. Disabling the idle timeout isn't really a good solution for safety reasons, but I'm open to hear other options. Maybe we can at least disable the part cooling fan during wait for filament change.

Filament runout is ideally something that is dealt with quickly, this also ensures best print quality :)

**Reply by Tiwatz** [reply]
I added this to the first section in ``macros.cfg``
``` 
[gcode_macro _CLIENT_VARIABLE]
#Set to 43200sec = 12h for pause macro
variable_idle_timeout     : 43200     ; time in sec until idle_timeout kicks in. Value 0 means that no value will be set or restored
```

**Comment by DougJoseph** [comment]
Without having read this, I posted a few minutes ago with an almost identical discussion with some proposals to at least talk about. :-)

**Comment by gg48gg** [comment]
Just wanted to say thank you.  I probably won't be back.  I am selling the 5M Pro and replaced it with another Bambu Lab P1 series.  The 5M Pro was promising and so was Klipper-mod, but I don't have the time to mess with it.

**Comment by MACCAMENZIES** [comment]
Just following on from this, I have noticed that when the filament runs out that the extruder moves to the very front left of the machine making it difficult to change the filament, IS there somewhere I can place a different x and y axis co-oridinate?

**Reply by DougJoseph** [reply]
There are variables for where it gets parked during that event. Quite some time back, some kind soul showed me where they were and I edited them and it worked. However, I just searched the discussion area and was unable to find that. I later had to redo my Klipper mod, forgot to make a backup, and lost that modification. I'd like to have it back.

I just looked inside the macros.cfg and the printer.cfg files, and in macros.cfg I found this:

```
variable_custom_park_x    : -105.0  ; custom x position; value must be within your defined min and max of X
variable_custom_park_y    : -105.0  ; custom y position; value must be within your defined min and max of Y

```

And I'm not positive, but it may be as simple as changing those from -105 to 105 (change neg. to pos.).

**Reply by MACCAMENZIES** [reply]
Worked a treat. Thanks Doug.

**Comment by MACCAMENZIES** [comment]
Thanks Doug. I'll give that a try

**Reply by DougJoseph** [reply]
Just circling back to say, it seems those are the correct variables to edit, as I edited them, and if memory serves (which is not a given) I think the next time I had a filament runout, the head parked at the back right instead of the front left.

**Comment by mstraciak** [comment]
I can live when the printer turns off the heating. i understand that there is safety reason.
but its not possible turn off the lost of home position?

**Reply by fernplant** [reply]
Its the timeout that does this. Its part of Klipper. AFAIK the only way to avoid it is to change the timeout to a longer value. I personally have setup notifications to my phone so that Im alerted when runout is detected or filament change needed so I can attend to it quickly

**Reply by DougJoseph** [reply]
Wow, that's neat. How do you setup phone notifications?

**Reply by fernplant** [reply]
Yes Ive found it a very handy feature. Sometimes smile when Im away from home and my watch tells me that my print is done. Its a feature of Moonraker. See docs here https://moonraker.readthedocs.io/en/latest/configuration/#notifier

**Reply by DougJoseph** [reply]
Awesome, thanks so much!

**Reply by DougJoseph** [reply]
@fernplant 
That was so easy, and by using my Discord server, it's free. So cool.

For others looking for quick and dirty example for this:

1. On my laptop, I opened my personal Discord server and added a new channel called #klipper 
2. In Discord app (again on my laptop), I right-clicked on my server, and navigated to Server Settings > Integrations
3. Clicked on Webhooks, and then "New Webhook"
4. I named the new webhook "Klipper Notifications" and set its channel to the #Klipper one I created above.
5. Save changes at the bottom, COPY the URL, and escape out. Paste the URL somewhere. You need it later. 
6. In Mainsail, I made note of my webcam's URL (with port # etc) for snapshots (stored in Settings > Webcams)
7. In Mainsail > Machine > moonraker.conf, I pasted the following, and did "Save and Restart"

**(Note: In the following, replace {WebhookID}/{WebhookToken} with the actual values you copied in step #5 above. Also, obviously, edit the webcam IP address with your own.)**
 
#moonraker.conf

[notifier print_start]
url: https://discord.com/api/webhooks/{WebhookID}/{WebhookToken}
events: started
body: Your printer started printing '{event_args[1].filename}'

[notifier print_complete]
url: https://discord.com/api/webhooks/{WebhookID}/{WebhookToken}
events: complete
body: Your printer completed printing '{event_args[1].filename}'
attach: http://192.168.1.24:8080?action=snapshot

[notifier print_error]
url: https://discord.com/api/webhooks/{WebhookID}/{WebhookToken}
events: error
body: {event_args[1].message}
attach: http://192.168.1.24:8080?action=snapshot

This worked a treat. The notification for a job completed looks like this:

In Discord app on my phone:
![IMG_9660](https://github.com/user-attachments/assets/4ac2c8df-e456-4b13-ae8b-7c2c3634d01e)

In Discord app on my laptop:
<img width="1728" alt="Screenshot 2024-11-24 at 1 37 11 AM" src="https://github.com/user-attachments/assets/ab40a259-24cb-46f6-8970-c78aa234512c">
